content/tutorials/build-apps/build-a-desktop-wallet-in-javascript.md

```diff
@@ -0,0 +1,615 @@
+
+This example shows how to do Inter Process Communication (IPC) in Electron. Technically, JavaScript has no true parallel processes or threading, because it follows a single-threaded event-driven paradigm. Nonetheless Electron provides us with two IPC modules called `ipcMain` and `ipcRenderer`. We can roughly equate those two to a backend process and a frontend process when we think in terms of client-server applications. It works as follows:
+
+1. Create a function that enables the frontend to subscribe to backend events (in `view/1_preload.js`)
+2. Make the function available by preloading it (webPreferences.preload during window creation)
+3. Create a frontend view
+4. Use that function in the frontend (e.g. 1_renderer.js, loaded in 1_hello.html) to attach a callback that handles frontend updates when the event is dispatched
+5. Dispatch the event from the backend (e.g. appWindow.webContents.send('update-ledger-index', value))
+
+In the package.json file we have already defined some prepared commands to run our application according to the steps comprising the
+structure of this tutorial. To get the application running at this early stage of development, run the following command:
+
+```console
+npm run hello
+```
+
+
+### 2.A. Show Ledger Updates by using WebSocket subscriptions
+
+**Full code for this step:** 
+[`2_async-subscribe.js`]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/desktop-js/2_async-subscribe.js),
+[`view/2_preload.js`]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/desktop-js/view/2_preload.js),
+[`view/2_async.html`]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/desktop-js/view/2_async.html),
+[`view/2_renderer.js`]({{target.github_forkurl}}/tree/{{target.github_branch}}/content/_code-samples/build-a-wallet/desktop-js/view/2_renderer.js).
+
+Our "Hello Ledger" application from Step 1. so far only shows the latest validated ledger sequence at the time when we opened it. So let's take things up a notch and add some dashboard like functionality where our wallet app will keep in sync with the ledger and display the latest specs and stats like a clock that is keeping track on time:
+
+![Screenshot: Step 2, show ledger updates](img/javascript-wallet-2.png)
+
+The code has been refactored (`1_hello.js` to `2_async-subscribe.js`) so that the main logic now resides in a main() function. This allows us to handle the application ready event by using an one-liner at the end of the code. We will do such refactorings regularly along our journey in order to keep the code well managed and readable.
```

Comment:
I'm not quite sure from this whether I should delete `1_hello.js` or how I should get the `2_async-subscribe.js` code. Please add explicit steps to copy code into the repo. 

(Is the expectation that I copy the code each time it's mentioned at the start of the section - or that I do it as we go through the instructions throughout?)

Reply:
I'd recommend showing the user exactly what's changing in each section rather than just having them copy the whole files from scratch because it's hard to tell what we're changing when the whole files change. (Ex. `2_preload.js` is identical to `1_preload.js` - so no change would actually have to happen.)

Having the files behind the scenes with the "snapshots" like `1_preload.js`, `2_preload.js`, etc. makes sense to me - but we should probably just be creating `preload.js` and updating in place as we go so that each change is explained clearly to the person working their way through the tutorial.

Reply:
@JST5000, We had this issue of flip-flopping from "we" to "you" before, this was because of the code-along style of the original python tutorial. As this one incorporates more code, with electron and the extra password step, somehow that explanatory style with "now we have" crept in. I guess I will reformat the description to stick to the code-along style, including descriptions where necessary. then all three wallet tutorials conform to the same style.

This will include working on just one file (e.g. preload.js instead of 38_preload.js) without the numbers, which will be used only for reference. While we're on it, look if theres a dusty box of medals somewhere in the basement, this one turned out way more complex than expected :)

Reply:
Obsolete, this has been rewritten completely